sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/executeImmediate.scala

```diff
@@ -0,0 +1,186 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import scala.util.{Either, Left, Right}
+
+import org.apache.spark.SparkException
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.expressions.{Alias, Expression, NamedExpression, VariableReference}
+import org.apache.spark.sql.catalyst.parser.{ParseException, ParserInterface}
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, SetVariable}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.{EXECUTE_IMMEDIATE, TreePattern}
+import org.apache.spark.sql.connector.catalog.CatalogManager
+import org.apache.spark.sql.errors.QueryCompilationErrors.unresolvedVariableError
+import org.apache.spark.sql.types.StringType
+
+/**
+ * Logical plan representing execute immediate query.
+ *
+ * @param args
+ *   parameters of query
+ * @param query
+ *   query string or variable
+ * @param targetVariables
+ *   variables to store the result of the query
+ */
+case class ExecuteImmediateQuery(
+  args: Seq[Expression],
+  query: Either[String, UnresolvedAttribute],
+  targetVariables: Option[Seq[UnresolvedAttribute]],
+  parser: ParserInterface)
+    extends UnresolvedLeafNode {
+  final override val nodePatterns: Seq[TreePattern] = Seq(EXECUTE_IMMEDIATE)
+}
+
+/**
+ * This rule substitutes execute immediate query node with plan that is passed as string literal
+ * or session parameter.
+ */
+class SubstituteExecuteImmediate(val catalogManager: CatalogManager)
+    extends Rule[LogicalPlan]
+    with ColumnResolutionHelper {
+
+  def resolveVariable(e: Expression): Expression = {
+
+    /**
+     * We know that the expression is either UnresolvedAttribute or Alias, as passed from the
+     * parser. If it is an UnresolvedAttribute, we look it up in the catalog and return it. If it
+     * is an Alias, we resolve the child and return an Alias with the same name.
+     */
+    e match {
+      case u: UnresolvedAttribute =>
+        getVariableReference(u.nameParts)
+      case a: Alias =>
+        Alias(resolveVariable(a.child), a.name)()
+
+      case other =>
+        throw SparkException.internalError(
+          "Unexpected variable expression in ParametrizedQuery: " + other)
```

Comment:
is it a user error? They can put an undefined variable in USING, right?

Reply:
parser needs to pass either Alias on UnresolvedAttribute here, so if we got here somehow, its a bug